### PR TITLE
12.0 fix chrome v8 crash moc

### PR DESCRIPTION
**Pull request by d-fence**
Since Chrome version 80, the V8 javascript engine is using an new
pointer compression feature [1]. This features tries to reserve more the
4GiB of memory. As a consequence, during HttpCase tests, odoo tries to
launch chrome headless in a subprocess but fails because of the
memory-limit-soft.

With this commit, the Chrome browser is spawned in a forked process
that removes the memory limit.

Also:
  * a new Chrome CLI switch is used to prevent crash reports to be sent
  to google
  * a bug that only occurs during Chrome crash is fixed [2]

[1] https://v8.dev/blog/v8-release-80
[2] None type object does not have a send method

### CODE REVIEW

File: odoo/tests/common.py [hunk 1]
```diff
@@ -537,8 +537,20 @@ def executable(self):
...
+    def _spawn_chrome(self):
+        pid = os.fork()
+        if pid != 0:
+            return pid
+        else:
+            if os.name == 'posix':
+                # prevent chrome crash because of huge memory reservation for V8
+                rlimit = resource.RLIMIT_RSS if platform.system() == 'Darwin' else resource.RLIMIT_AS
+                soft, hard = resource.getrlimit(rlimit)
+                resource.setrlimit(rlimit, (resource.RLIM_INFINITY, resource.RLIM_INFINITY))
```
Comment (xmo-odoo) [inline odoo/tests/common.py]: can use `os.dup2(<new>, sys.stderr.fileno())` in order to "replace" the inherited stderr by something else.

File: odoo/tests/common.py [hunk 2]
```diff
@@ -537,8 +537,20 @@ def executable(self):
...
         raise unittest.SkipTest("Chrome executable not found")
 
+    def _spawn_chrome(self):
+        pid = os.fork()
+        if pid != 0:
+            return pid
+        else:
+            if os.name == 'posix':
+                # prevent chrome crash because of huge memory reservation for V8
+                rlimit = resource.RLIMIT_RSS if platform.system() == 'Darwin' else resource.RLIMIT_AS
```
Comment (xmo-odoo) [inline odoo/tests/common.py]: Can skip everything for darwin: issue is when we're rlimiting vmem because pointer compression feature performs a large reservation upfront, but we're not rlimiting vmem on OSX because the OS almost always does a huge reservation upfront for all processes anyway, used to be processes got ~2.4GB VM by default and I think it's higher now (it's not rare to have aggregate VM sizes in the 200~300GB ranges). So the issue should not occur there.

File: odoo/tests/common.py [hunk 3]
```diff
@@ -558,17 +570,18 @@ def _chrome_start(self):
             '--remote-debugging-address': HOST,
             '--remote-debugging-port': str(self.devtools_port),
             '--no-sandbox': '',
+            '--disable-crash-reporter': '',
+            '--disable-gpu': '',
         }
-        cmd = [self.executable]
-        cmd += ['%s=%s' % (k, v) if v else k for k, v in switches.items()]
+        self.cmd = [self.executable]
+        self.cmd += ['%s=%s' % (k, v) if v else k for k, v in switches.items()]
```
[the remark this answers is not in the document]
Comment (d-fence) [reply odoo/tests/common.py]: I prefer spawn_chrome as it's the only method to spawn Chrome, disabling hard limit or not.

File: odoo/tests/common.py [hunk 4]
```diff
@@ -537,8 +537,20 @@ def executable(self):
...
 
+    def _spawn_chrome(self):
+        pid = os.fork()
+        if pid != 0:
+            return pid
+        else:
+            if os.name == 'posix':
+                # prevent chrome crash because of huge memory reservation for V8
+                rlimit = resource.RLIMIT_RSS if platform.system() == 'Darwin' else resource.RLIMIT_AS
+                soft, hard = resource.getrlimit(rlimit)
```
[the remark this answers is not in the document]
Comment (d-fence) [reply odoo/tests/common.py]: That was my first idea, but I think that it's better to spawn an independent chrome process.

File: odoo/tests/common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (odony) [reply odoo/tests/common.py]: I agree for the instance attribute, and how about inline the contents of `spawn_chrome()` in `_chrome_start` for simplicity?

File: odoo/tests/common.py (hunk 2 above)
Comment (odony) [reply odoo/tests/common.py]: I meant explicit Python comments, in addition to the commit message that should have the same info

File: odoo/tests/common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (d-fence) [reply odoo/tests/common.py]: I would prefer to keep the fork thing in a separate method, for clarity.

File: odoo/tests/common.py [hunk 5]
```diff
@@ -537,8 +537,24 @@ def executable(self):
...
+        else:
+            if os.name == 'posix' and platform.system() != 'Darwin':
+                # since the introduction of pointer compression in Chrome 80 (v8 v8.0),
+                # the memory reservation algorithm requires more than 8GiB of virtual mem for alignment
+                # this exceeds our default memory limits.
+                # OSX already reserve huge memory for processes
+                resource.setrlimit(resource.RLIMIT_AS, (resource.RLIM_INFINITY, resource.RLIM_INFINITY))
+                stdout_replacement = os.open(os.devnull, os.O_WRONLY, 0)
+                os.dup2(stdout_replacement, sys.stderr.fileno())
+                os.close(stdout_replacement)
```
Comment (xmo-odoo) [inline odoo/tests/common.py]: A slightly cleaner version might be
```python
with contextlib.closing(os.open(os.devnull, ...)) as stdout_replacement:
    os.dup2(stdout_replacement, sys.stderr.fileno())
```
or
```python
with open(os.devnull, 'w') as stdout_replament:
    os.dup2(stdout_replacement.fileno(), sys.stderr.fileno())
```

Comment (d-fence) [reply odoo/tests/common.py]: I agree and prefer the second one

**Comment by d-fence** [comment]
@odony about the target version, even if it's only  developer convenience, I think it could be quite annoying for the bugfix team. When it happens, it's not obvious that Chrome crashed, its stderr is lost in /dev/null.

File: odoo/tests/common.py [hunk 6]
```diff
@@ -537,8 +537,23 @@ def executable(self):
...
+            return pid
+        else:
+            if os.name == 'posix' and platform.system() != 'Darwin':
+                # since the introduction of pointer compression in Chrome 80 (v8 v8.0),
+                # the memory reservation algorithm requires more than 8GiB of virtual mem for alignment
+                # this exceeds our default memory limits.
+                # OSX already reserve huge memory for processes
+                resource.setrlimit(resource.RLIMIT_AS, (resource.RLIM_INFINITY, resource.RLIM_INFINITY))
+                with open(os.devnull, 'wb', 0) as stdout_replacement:
+                    os.dup2(stdout_replacement.fileno(), sys.stderr.fileno())
```
Comment (odony) [inline odoo/tests/common.py]: - The dup2 is also needed on Darwin, I guess
- Any reason to call it `stdout_replacement` and then use it for stderr only? ;-)

File: odoo/tests/common.py [hunk 7]
```diff
@@ -622,6 +637,8 @@ def _websocket_send(self, method, params=None):
         """
         send chrome devtools protocol commands through websocket
         """
+        if self.ws is None:
```
Comment (odony) [inline odoo/tests/common.py]: perhaps this fix belongs in a separate commit, to keep things clear?

File: odoo/tests/common.py [hunk 8]
```diff
@@ -537,8 +537,23 @@ def executable(self):
 
         raise unittest.SkipTest("Chrome executable not found")
 
+    def _spawn_chrome(self, cmd):
+        pid = os.fork()
```
Comment (odony) [inline odoo/tests/common.py]: `fork()` won't work on Windows but I guess it's fine because `executable()` will raise first. For consistency we might still want to move the posix test at the top and return early for non-posix. What do you think?

**Comment by odony** [comment]
> @odony about the target version, even if it's only developer convenience, I think it could be quite annoying for the bugfix team. When it happens, it's not obvious that Chrome crashed, its stderr is lost in /dev/null.

Hmm, as long as it only impacts the testing infrastructure I don't see any significant risk for stable.

Unless @KangOl has any objection? ;)

File: odoo/tests/common.py (hunk 6 above)
Comment (d-fence) [reply odoo/tests/common.py]: good catch, I tested with stdout :smiley: 
And yes I have to separate the Darwin stuff

File: odoo/tests/common.py (hunk 7 above)
Comment (d-fence) [reply odoo/tests/common.py]: yep

File: odoo/tests/common.py [hunk 9]
```diff
@@ -558,17 +572,18 @@ def _chrome_start(self):
             '--remote-debugging-address': HOST,
             '--remote-debugging-port': str(self.devtools_port),
             '--no-sandbox': '',
+            '--disable-crash-reporter': '',
```
Comment (KangOl) [inline odoo/tests/common.py]: Does this flag also exists in Chromium?

File: odoo/tests/common.py (hunk 8 above)
Comment (d-fence) [reply odoo/tests/common.py]: I agree

File: odoo/tests/common.py (hunk 6 above)
Comment (KangOl) [reply odoo/tests/common.py]: Yes, dup2 should be called on all platforms.

File: odoo/tests/common.py [hunk 10]
```diff
@@ -498,13 +500,11 @@ def signal_handler(self, sig, frame):
...
-            self._logger.info("Closing chrome headless with pid %s", self.chrome_process.pid)
+        if self.chrome_pid is not None:
+            self._logger.info("Closing chrome headless with pid %s", self.chrome_pid)
             self._websocket_send('Browser.close')
-            if self.chrome_process.poll() is None:
-                self._logger.info("Terminating chrome headless with pid %s", self.chrome_process.pid)
-                self.chrome_process.terminate()
-                self.chrome_process.wait()
+            self._logger.info("Terminating chrome headless with pid %s", self.chrome_pid)
+            os.kill(self.chrome_pid, 15)
```
Comment (KangOl) [inline odoo/tests/common.py]: use `signal.SIGTERM`

File: odoo/tests/common.py (hunk 9 above)
Comment (d-fence) [reply odoo/tests/common.py]: Yep [here](https://cs.chromium.org/chromium/src/headless/app/headless_shell_switches.cc?q=kDisableCrashReporter&sq=package:chromium&type=cs&l=32)

File: odoo/tests/common.py (hunk 10 above)
Comment (d-fence) [reply odoo/tests/common.py]: ok

**Comment by d-fence** [comment]
@odony I think that this one can be r-plused, no ?

**Comment by d-fence** [comment]
@odony thanks
